tidb-binlog/driver/reader/reader.go

```diff
@@ -145,88 +131,78 @@ func NewReader(cfg *Config) (r *Reader, err error) {
 // Close shuts down the reader
 func (r *Reader) Close() {
 	close(r.stop)
-
-	r.client.Close()
 }
 
 // Messages returns a chan that contains unread buffered message
 func (r *Reader) Messages() (msgs <-chan *Message) {
 	return r.msgs
 }
 
-func (r *Reader) getOffsetByTS(ts int64, conf *sarama.Config) (offset int64, err error) {
-	// set true to retrive error
-	conf.Consumer.Return.Errors = true
-	seeker, err := NewKafkaSeeker(r.cfg.KafkaAddr, conf)
-	if err != nil {
-		err = errors.Trace(err)
-		return
-	}
-
-	topic, partition := r.getTopic()
-	log.Debug("get offset",
-		zap.String("topic", topic),
-		zap.Int32("partition", partition),
-		zap.Int64("ts", ts))
-	offsets, err := seeker.Seek(topic, ts, []int32{partition})
+func (r *Reader) getOffsetByTS(ts int64, topic string, partitions []int32) (offset int64, err error) {
+	offsets, err := r.client.SeekOffsetFromTS(ts, topic, partitions)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
-
 	offset = offsets[0]
-
 	return
 }
 
 func (r *Reader) run() {
 	offset := r.cfg.Offset
 	log.Debug("start at", zap.Int64("offset", offset))
+	consumerChan := make(chan *kafka_consumer.KafkaMsg, r.cfg.getMessageBufferSize())
+	done := make(chan struct{})
 
-	consumer, err := sarama.NewConsumerFromClient(r.client)
-	if err != nil {
-		log.Fatal("create kafka consumer failed", zap.Error(err))
-	}
-	defer consumer.Close()
-	topic, partition := r.getTopic()
-	partitionConsumer, err := consumer.ConsumePartition(topic, partition, offset)
-	if err != nil {
-		log.Fatal("create kafka partition consumer failed", zap.Error(err))
-	}
-	defer partitionConsumer.Close()
-
-	// add select to avoid message blocking while reading
-	for {
-		select {
-		case <-r.stop:
-			// clean environment
-			partitionConsumer.Close()
-			close(r.msgs)
-			log.Info("reader stop to run")
-			return
-		case kmsg := <-partitionConsumer.Messages():
-			log.Debug("get kafka message", zap.Int64("offset", kmsg.Offset))
-			binlog := new(pb.Binlog)
-			err := binlog.Unmarshal(kmsg.Value)
-			if err != nil {
-				log.Warn("unmarshal binlog failed", zap.Error(err))
-				continue
-			}
-			if r.cfg.CommitTS > 0 && binlog.CommitTs <= r.cfg.CommitTS {
-				log.Warn("skip binlog CommitTs", zap.Int64("commitTS", binlog.CommitTs))
-				continue
-			}
-
-			msg := &Message{
-				Binlog: binlog,
-				Offset: kmsg.Offset,
-			}
+	go func() {
+		// add select to avoid message blocking while reading
+		for {
 			select {
-			case r.msgs <- msg:
 			case <-r.stop:
-				// In the next iteration, the <-r.stop would match again and prepare to quit
-				continue
+				// clean environment
+				r.client.Close()
+				close(r.msgs)
+				close(done)
```

Comment:
if it handle client exiting logic in `r.client.Close()`, do we still need `done` channel?

Reply:
`r.client.Close()` means release client resource
close `done` channel  means stop consume, but do not release resource, maybe I should change this name